src/generator/mod.rs

```diff
@@ -0,0 +1,149 @@
+//! Source generator module.
+
+use vhdl::{Component, Mode, Port, Type};
+
+use crate::{
+    phys::{BitField, Dir, Stream},
+    Streamlet,
+};
+use crate::generator::vhdl::{Record};
+
+pub mod vhdl;
+
+fn log2ceil(v: usize) -> usize { (v as f64).log2().ceil() as usize }
+
+impl From<BitField> for Type {
+    fn from(b: BitField) -> Self {
+        if b.width() == 1 {
+            Type::Bit
+        } else {
+            Type::BitVec {
+                width: b.width()
+            }
+        }
+    }
+}
+
+impl From<Dir> for Mode {
+    fn from(d: Dir) -> Self {
+        if d == Dir::Downstream {
```

Comment:
You could write this using a `match`:
```rust
match d {
    Dir::Downstream => Mode::Out,
    Dir::Upstream => Mode::In,
}
```

The benefit here is that someone can directly see that `Dir` has only two variants.